internal/server/pool_test.go

```diff
@@ -53,7 +54,455 @@ func TestPoolSpike(t *testing.T) {
 
 	be.Equal(rt, 5, p.Stats().Idle)
-	be.NilErr(rt, p.Close())
+}
+
+func TestMaxOpenConnsUpdate(t *testing.T) {
+	t.Run("basic", func(t *testing.T) {
+		cfg := genBasePoolConfig()
+		cfg.MaxOpen = 3
+		cfg.MaxIdle = 2
+
+		rt := be.Relaxed(t)
+
+		p, err := NewPool(cfg)
+		be.NilErr(rt, err)
+		defer func() { be.NilErr(rt, p.Close()) }()
+
+		conn0, err := p.AcquireConn()
+		be.NilErr(rt, err)
+
+		conn1, err := p.AcquireConn()
+		be.NilErr(rt, err)
+
+		conn2, err := p.AcquireConn()
+		be.NilErr(rt, err)
+
+		be.Equal(t, 3, p.Stats().OpenConnections)
+
+		p.SetMaxOpenConns(2)
+
+		p.ReleaseConn(conn0)
+		p.ReleaseConn(conn1)
+
+		// Ensure that extra connections are not added back.
+		be.Equal(t, 2, p.Stats().OpenConnections)
+
+		p.ReleaseConn(conn2)
+
+		be.Equal(t, 2, p.Stats().OpenConnections)
+	})
+
+	t.Run("maxOpen < maxIdle", func(t *testing.T) {
+		cfg := genBasePoolConfig()
+		cfg.MaxOpen = 3
+		cfg.MaxIdle = 3
+
+		rt := be.Relaxed(t)
+
+		p, err := NewPool(cfg)
+		be.NilErr(rt, err)
+		defer func() { be.NilErr(rt, p.Close()) }()
+
+		conn0, err := p.AcquireConn()
+		be.NilErr(rt, err)
+
+		conn1, err := p.AcquireConn()
+		be.NilErr(rt, err)
+
+		conn2, err := p.AcquireConn()
+		be.NilErr(rt, err)
+
+		be.Equal(t, 3, p.Stats().OpenConnections)
+		be.Equal(t, 0, p.Stats().Idle)
+
+		p.ReleaseConn(conn0)
+		p.ReleaseConn(conn1)
+		p.ReleaseConn(conn2)
+
+		be.Equal(t, 3, p.Stats().OpenConnections)
+		be.Equal(t, 3, p.Stats().Idle)
+
+		p.SetMaxOpenConns(2)
+
+		be.Equal(t, 2, p.Stats().OpenConnections)
+		be.Equal(t, 2, p.Stats().Idle)
+	})
+}
+
+func TestMaxIdleConnsUpdate(t *testing.T) {
+	t.Run("basic", func(t *testing.T) {
+		cfg := genBasePoolConfig()
+		cfg.MaxOpen = 3
+		cfg.MaxIdle = 3
+
+		rt := be.Relaxed(t)
+
+		p, err := NewPool(cfg)
+		be.NilErr(rt, err)
+		defer func() { be.NilErr(rt, p.Close()) }()
+
+		conn0, err := p.AcquireConn()
+		be.NilErr(rt, err)
+
+		conn1, err := p.AcquireConn()
+		be.NilErr(rt, err)
+
+		conn2, err := p.AcquireConn()
+		be.NilErr(rt, err)
+
+		be.Equal(t, 3, p.Stats().OpenConnections)
+		be.Equal(t, 0, p.Stats().Idle)
+
+		p.ReleaseConn(conn0)
+		p.ReleaseConn(conn1)
+		p.ReleaseConn(conn2)
+
+		be.Equal(t, 3, p.Stats().OpenConnections)
+		be.Equal(t, 3, p.Stats().Idle)
+
+		p.SetMaxIdleConns(1)
+
+		be.Equal(t, 1, p.Stats().OpenConnections)
+		be.Equal(t, 1, p.Stats().Idle)
+	})
+
+	t.Run("maxIdle > maxOpen", func(t *testing.T) {
+		cfg := genBasePoolConfig()
+		cfg.MaxOpen = 2
+		cfg.MaxIdle = 2
+
+		rt := be.Relaxed(t)
+
+		p, err := NewPool(cfg)
+		be.NilErr(rt, err)
+		defer func() { be.NilErr(rt, p.Close()) }()
+
+		conn0, err := p.AcquireConn()
+		be.NilErr(rt, err)
+
+		conn1, err := p.AcquireConn()
+		be.NilErr(rt, err)
+
+		be.Equal(t, 2, p.Stats().OpenConnections)
+		be.Equal(t, 0, p.Stats().Idle)
+
+		p.ReleaseConn(conn0)
+		p.ReleaseConn(conn1)
+
+		be.Equal(t, 2, p.Stats().OpenConnections)
+		be.Equal(t, 2, p.Stats().Idle)
+
+		p.SetMaxIdleConns(5)
+
+		be.Equal(t, 2, p.maxIdle)
```

Comment:
Do we want to check here also the `p.Stats{}.OpenConnections` and `p.Stats().Idle`?

Reply:
We could, but we are already testing that in the previous sub-test. In this test, our main target is to confirm that maxIdle does not go above `maxOpen` value.